### PR TITLE
Rework of the library

### PR DESCRIPTION
**Pull request by bensherred**
## Important Information

I recommend bumping this package to a new major release. This library has been rewritten and all classes are different. It will require users to update their implementation.

The required PHP version has also been bumped to PHP 7.2 to support the latest versions of composer dependencies

## Changelog

This PR continues on from and extends #9 which has been closed due to inactivity.

- Seperated the tests into individual test classes
- Removed hhvm from Travis CI config
- Set the schema to always use HTTPS
- Changed the way each request is handled (you can refer to the [wiki](https://github.com/TruckersMP/API-Client/wiki) for more information
- We now have a model to represent each API endpoint including new properties
- Simplified the README.md
- Added extended documentation via the [wiki](https://github.com/TruckersMP/API-Client/wiki)
- Remvoed the examples from the package (the [wiki](https://github.com/TruckersMP/API-Client/wiki) has plenty of examples)
- Updated composer dependencies to the latest version
- Set the minimum PHP version to PHP 7.2
- Added PHP 7.2 to the Travis CI config

## Future Development

If an API endpoint gets updated in the future, you can easily update the package by adding the new properties to the model or creating a new model all together.

You can also chain requests together like the company() request.

### CODE REVIEW

File: src/Collections/CompanyCollection.php
Comment (ShawnCZek) [inline src/Collections/CompanyCollection.php]: The documentation is not accurate.

File: src/Collections/PostsCollection.php
Comment (ShawnCZek) [inline src/Collections/PostsCollection.php]: The documentation is not accurate.

File: src/Collections/RoleCollection.php
Comment (ShawnCZek) [inline src/Collections/RoleCollection.php]: The documentation is not accurate.

File: src/Collections/ServerCollection.php
Comment (ShawnCZek) [inline src/Collections/ServerCollection.php]: The documentation is not accurate.

File: src/Models/CompanyIndex.php
Comment (ShawnCZek) [inline src/Models/CompanyIndex.php]: The documentation is not updated.

File: src/Models/CompanyRole.php
Comment (ShawnCZek) [inline src/Models/CompanyRole.php]: This comment does not seem to be correct. I suggest these changes:

```suggestion
     * The owner of the company.
```

File: src/Models/Player.php
Comment (ShawnCZek) [inline src/Models/Player.php]: A little typo.

```suggestion
     * URL to the small avatar on the website.
```

File: src/Models/Player.php
Comment (ShawnCZek) [inline src/Models/Player.php]: The bans should probably be fetched once the getter is used. Bans are not that important in some cases (especially when some fields are already included in the player model). Imagine fetching a huge list of members. It would request bans for each of them.

File: src/Models/Player.php
Comment (ShawnCZek) [inline src/Models/Player.php]: Use getter instead.

File: src/Models/Player.php
Comment (ShawnCZek) [inline src/Models/Player.php]: Use getter instead.

File: src/Models/Player.php
Comment (ShawnCZek) [inline src/Models/Player.php]: Use getter instead.

File: src/Requests/Companies/MemberRequest.php
Comment (ShawnCZek) [inline src/Requests/Companies/MemberRequest.php]: A typo in this comment.

```suggestion
     * The ID of the requested member.
```

File: src/Requests/Companies/PostRequest.php
Comment (ShawnCZek) [inline src/Requests/Companies/PostRequest.php]: A typo in the comment.

```suggestion
     * The ID of the requested post.
```

File: src/Requests/CompanyRequest.php
Comment (ShawnCZek) [inline src/Requests/CompanyRequest.php]: It seems like this variable is not used.

File: .gitignore
Comment (ShawnCZek) [inline .gitignore:9]: Weirdly enough, I get `.phpunit.result.cache` generated instead of a new folder. Can you add it here, too, please?

**Review by ShawnCZek — CHANGES_REQUESTED** [review]
Just a small changes must be done. Otherwise, it looks like it is working fine (according to the tests that were thankfully created).

**Comment by ShawnCZek** [comment]
Also, it seems that the Composer lock file has not been updated:

```
Warning: The lock file is not up to date with the latest changes in composer.json. You may be getting outdated dependencies. Run update to update them.
```

Can you do so, please?

File: src/Collections/CompanyCollection.php
Comment (bensherred) [reply src/Collections/CompanyCollection.php]: Updated description to `Create a new CompanyCollection instance.`

File: src/Collections/PostsCollection.php
Comment (bensherred) [reply src/Collections/PostsCollection.php]: Updated description to `Create a new PostsCollection instance.`

File: src/Collections/RoleCollection.php
Comment (bensherred) [reply src/Collections/RoleCollection.php]: Updated description to `Create a new RoleCollection instance.`

File: src/Collections/ServerCollection.php
Comment (bensherred) [reply src/Collections/ServerCollection.php]: Updated description to `Create a new ServerCollection instance.`

File: src/Models/Social.php
Comment (ShawnCZek) [inline src/Models/Social.php]: Shouldn't the construct rather accept parameters of social sites instead of an array? Like the model `Game`.

File: src/Requests/Companies/PostRequest.php
Comment (ShawnCZek) [inline src/Requests/Companies/PostRequest.php]: `$postId` should be probably `int`, too. Also, do not forget to update the documentation.

File: src/Models/Checksum.php
Comment (ShawnCZek) [inline src/Models/Checksum.php]: Shouldn't the construct rather accept parameters of social sites instead of an array? Like the model `Game`.

File: src/Client.php
Comment (ShawnCZek) [inline src/Client.php:92]: Are you planning to change the method name to match it with the request class? Or will you keep it like this?

File: src/Models/Company.php
Comment (ShawnCZek) [inline src/Models/Company.php:200]: I would suggest adding constants for this attribute which are `'Close'` and `'Open'`. Then we could also add methods to check whether the recruitment is closed or not.

File: src/Models/Player.php
Comment (ShawnCZek) [inline src/Models/Player.php]: Can be `null`.

```suggestion
     * @var \Carbon\Carbon|null
```

File: src/Requests/Company/MemberIndexRequest.php
Comment (ShawnCZek) [inline src/Requests/Company/MemberIndexRequest.php]: Not accurate documentation.

```suggestion
     * Create a new MemberIndexRequest instance.
```

File: src/Requests/Company/PostIndexRequest.php
Comment (ShawnCZek) [inline src/Requests/Company/PostIndexRequest.php]: Not accurate documentation.

```suggestion
     * Create a new PostIndexRequest instance.
```

File: src/Requests/Company/RoleIndexRequest.php
Comment (ShawnCZek) [inline src/Requests/Company/RoleIndexRequest.php]: Not accurate documentation.

```suggestion
     * Create a new RoleIndexRequest instance.
```

File: src/Requests/Company/RoleRequest.php
Comment (ShawnCZek) [inline src/Requests/Company/RoleRequest.php]: Not accurate documentation.

```suggestion
     * Create a new RoleRequest instance.
```

File: tests/Unit/CompanyTest.php
Comment (ShawnCZek) [inline tests/Unit/CompanyTest.php]: Looking at the tests now, this seems to be useless as this assertation happens only when the social information is `null`.

**Review by ShawnCZek — CHANGES_REQUESTED** [review]

File: src/Requests/Company/PostRequest.php
Comment (ShawnCZek) [inline src/Requests/Company/PostRequest.php]: The variable is missing a data type.

```suggestion
     * @param int $postId
```

File: src/Client.php
Comment (bensherred) [reply src/Client.php:92]: I think leaving it as `companies()` makes the most sense

File: tests/Unit/CompanyTest.php
Comment (bensherred) [reply tests/Unit/CompanyTest.php]: Gone ahead and removed asserting if `null`, also added a check to make sure `getSocial()` returns an instance of the `Social` model

File: tests/Unit/CompanyTest.php
Comment (ShawnCZek) [inline tests/Unit/CompanyTest.php]: This is also checked at the beginning of this method.

Comment (bensherred) [reply tests/Unit/CompanyTest.php]: Oops, will go ahead and remove

**Review by ShawnCZek — APPROVED** [review]
Everything seems to be alright now. Just let me know whether you want to publish this with finished documentation, or we can merge it as soon as we get other approvals.

Also, once this gets merged to `develop`, the license along with the composer files will be updated to hold the most recent information about all contributors correctly.

**Comment by bensherred** [comment]
> Everything seems to be alright now. Just let me know whether you want to publish this with finished documentation, or we can merge it as soon as we get other approvals.

As it's a fairly big change, I think it would be best to have the documentation to go with it. I will try and finish the core documentation tonight 😃

**Comment by bensherred** [comment]
Sorry it's taken a bit longer than expected, but I believe all the initial documentation has been added to the Wiki. Happy to tweek things as things come up, but I think there is enough to publish a new release.

**Pull request merged**
